src/main/java/org/opensearch/neuralsearch/highlight/NeuralHighlighterEngine.java

```diff
@@ -0,0 +1,306 @@
+     * @return The highlighting results
+     */
+    public List<Map<String, Object>> fetchModelResults(String modelId, String question, String context) {
+        PlainActionFuture<List<Map<String, Object>>> future = PlainActionFuture.newFuture();
+
+        SentenceHighlightingRequest request = SentenceHighlightingRequest.builder()
+            .modelId(modelId)
+            .question(question)
+            .context(context)
+            .build();
+
+        mlCommonsClient.inferenceSentenceHighlighting(request, future);
+
+        try {
+            return future.actionGet();
+        } catch (Exception e) {
+            throw new OpenSearchException(
+                String.format(Locale.ROOT, "Error during sentence highlighting inference from model [%s]", modelId),
+                e
+            );
+        }
+    }
+
+    /**
+     * Applies highlighting to the original context based on the ML model response
+     *
+     * @param context The original document text
+     * @param highlightResults The highlighting results from the ML model
+     * @return Formatted text with highlighting
+     */
+    public String applyHighlighting(String context, List<Map<String, Object>> highlightResults) {
+        List<int[]> validHighlights = new ArrayList<>();
+
+        if (highlightResults != null && !highlightResults.isEmpty()) {
+            Map<String, Object> result = highlightResults.getFirst();
+            if (result != null) {
+                // Get the "highlights" list from the result
+                Object highlightsObj = result.get(MODEL_INFERENCE_RESULT_KEY);
+
+                if (highlightsObj instanceof List<?> highlightsList) {
+                    for (Object item : highlightsList) {
+                        if (item instanceof Map<?, ?> map) {
+                            Map<String, Object> safeMap = new java.util.HashMap<>();
+                            for (Map.Entry<?, ?> entry : map.entrySet()) {
+                                safeMap.put(entry.getKey().toString(), entry.getValue());
+                            }
+
+                            // Extract and validate positions
+                            Object startObj = safeMap.get(MODEL_INFERENCE_RESULT_START_KEY);
+                            Object endObj = safeMap.get(MODEL_INFERENCE_RESULT_END_KEY);
+
+                            int[] positions = validateHighlightPositions(startObj, endObj, context.length());
+                            if (positions != null) {
+                                validHighlights.add(positions);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        // If no valid highlights, return the original text
+        if (validHighlights.isEmpty()) {
+            return context;
+        }
+
+        // Sort highlights by start position (ascending)
+        validHighlights.sort(Comparator.comparingInt(pos -> pos[0]));
```

Comment:
model does not guarantee to return sorted value? Shouldn't we enforce it just by throwing an error when value is not sorted while adding tags?

Reply:
We'll have sorted response from our pre-trained model. But we cannot make sure customized model have sorted response. But I still agree the solution to make it simpler handle in client side(neural-search plugin) with exception while it has unsorted response.

Reply:
Yes. The main concern is a latency due to this sorting.